### PR TITLE
rest-dhcpd: initial release

### PR DESCRIPTION
**Pull request by Seitanas**
Initial release of REST-DHCPD.

@vinted/sre

### CODE REVIEW

File: pkg/configdb/configdb.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"encoding/json"
+	"log"
+	"os"
+	"path"
+	"sync"
+)
+
+var mu sync.Mutex
```
Comment (GiedriusS) [inline pkg/configdb/configdb.go]: Please remove this and other global variables: https://github.com/uber-go/guide/blob/master/style.md#reduce-scope-of-variables, https://github.com/uber-go/guide/blob/master/style.md#avoid-mutable-globals.

File: pkg/configdb/configdb.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	AuthToken          string
+	ListenInterface    string
+	HTTPListenAddress  string
+	TLSEnabled         bool
+	TLSPrivateKeyFile  string
+	TLSCertificateFile string
+	Options            interface{}
+}
+
+type Clients struct {
```
Comment (GiedriusS) [inline pkg/configdb/configdb.go:28]: Maybe a simple `type Clients []Client` would suffice?

File: pkg/configdb/configdb.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	Hostname string
+	MAC      string
+	IP       string
+	Options  interface{}
+}
+
+var DB *Clients
+var Config *GlobalOptions
+
+func Init(config_path string) error {
```
Comment (GiedriusS) [inline pkg/configdb/configdb.go]: Variable names usually use mixed caps as per the Effective Go book: https://go.dev/doc/effective_go#mixed-caps. I suggest taking a look at it.

File: pkg/configdb/configdb.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	content, err := os.ReadFile(path.Join(config_path, "rest-dhcpd-clients.json"))
+	if err != nil {
+		log.Printf("%s", err)
+		return err
```
Comment (GiedriusS) [inline pkg/configdb/configdb.go:45]: Please don't `log` all over the place but prefer to return wrapped errors with context. In this case, I suggest doing: `return fmt.Errorf("reading %s: %w", path.Join(config_path, "rest-dhcpd-clients.json"), err)`.You can put the full path into a separate variable to facilitate re-using the same formatting logic.

File: pkg/configdb/configdb.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+func Save() {
+	content, err := json.MarshalIndent(DB, "", "  ")
+	if err != nil {
+		log.Printf("%s", err)
+	}
+	mu.Lock()
+	err = os.WriteFile("rest-dhcp-clients.json", content, 0644)
+	mu.Unlock()
+	if err != nil {
+		log.Printf("%s", err)
```
Comment (GiedriusS) [inline pkg/configdb/configdb.go]: Please return `error`from this function and handle it in the caller.

File: pkg/configdb/configdb.go (hunk 1 above)
Comment (Seitanas) [reply pkg/configdb/configdb.go]: @GiedriusS fixed

File: pkg/configdb/configdb.go (hunk 2 above)
Comment (Seitanas) [reply pkg/configdb/configdb.go:28]: Didn't came up with easier solution, unfortunatlely.

File: pkg/configdb/configdb.go (hunk 3 above)
Comment (Seitanas) [reply pkg/configdb/configdb.go]: Fixed

File: pkg/configdb/configdb.go (hunk 4 above)
Comment (Seitanas) [reply pkg/configdb/configdb.go:45]: Fixed

File: pkg/configdb/configdb.go (hunk 5 above)
Comment (Seitanas) [reply pkg/configdb/configdb.go]: fixed

File: pkg/rest/rest.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+	"log"
+	"net/http"
+	//"sync"
+)
+
+func index(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	fmt.Fprintf(w, "<html><body>REST-DHCPD<br>https://github.com/vinted/rest-dhcpd</body></html>")
+}
+
+func client_config_show(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
```
Comment (lukassup) [inline pkg/rest/rest.go]: I think `camelCase` is needed by Go conventions to make linters happy in the future

Comment (Seitanas) [reply pkg/rest/rest.go]: @lukassup fixed

Comment (lukassup) [reply pkg/rest/rest.go]: thank you!

File: pkg/rest/rest.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	if id == -1 {
+		fmt.Fprintf(w, "No configuration to delete for: %s\n", mac)
+	} else {
+		deleteClientConfig(id)
+		fmt.Fprintf(w, "Deleting config for MAC: %s, ID: %d\n", mac, id)
+	}
+
+}
+
+func clientConfigShow_full(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
```
Comment (lukassup) [inline pkg/rest/rest.go]: looks like a leftover after camelCase refactor

File: pkg/dhcpd/dhcpd.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+func StartServer() {
+	lease, _ := time.ParseDuration(strconv.Itoa(configdb.Config.LeaseDuration) + "s")
+	handler := &DHCPHandler{
+		ip:            net.ParseIP(configdb.Config.IP),
+		leaseDuration: lease,
+		options:       BuildOptions(configdb.Config.Options),
+	}
+	cn, err := conn.NewUDP4BoundListener(configdb.Config.ListenInterface, ":67")
+	if err != nil {
+		log.Fatal("NewUDP4BoundListener")
```
Comment (lukassup) [inline pkg/dhcpd/dhcpd.go]: We may want a better error message here (from `err`). Would be nice to know if port is already in use or something else

File: pkg/rest/rest.go (hunk 7 above)
Comment (Seitanas) [reply pkg/rest/rest.go]: fixed

File: pkg/dhcpd/dhcpd.go (hunk 8 above)
Comment (Seitanas) [reply pkg/dhcpd/dhcpd.go]: @lukassup fixed

**Review by lukassup — APPROVED** [review]

**Pull request merged**